### PR TITLE
Discard non-close messages in close timeout window

### PR DESCRIPTION
**Pull request by lenard-mosys**
## What do these changes do?

Waits for a close message in the close timeout window, not just the first message.

## Are there changes in behavior for the user?

Previously spurious abnormally closed websocket connections could be reported to the server, even if the client responded to the close message with a close in a timely manner.

## Is it a substantial burden for the maintainers to support this?

No

## Related issue number

https://github.com/aio-libs/aiohttp/issues/9506

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] ~~Documentation reflects the changes~~
  - It is a bugfix.
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
- [x] Add a new news fragment into the `CHANGES/` folder

### CODE REVIEW

**Comment by lenard-mosys** [comment]
Breaks this test:

https://github.com/aio-libs/aiohttp/blob/9e975d275ec6421a0d7d1649dacfb0ebf6ff72bf/tests/test_web_websocket_functional.py#L1235-L1254

IMO the test is broken, or at least names and comments are misleading. I'm a bit perplexed why this test even passed before. Note that the test tests for the close code of the `ClientWebSocketResponse`, not for the server's `WebSocketResponse`. The close code for `ws` within the `handler` should indeed be abnormal closure, but that's never tested.

I'm not entirely sure what this test intends to test, or even if it intends to test server or client side behavior.

**Comment by bdraco** [comment]
> I'm not entirely sure what this test intends to test, or even if it intends to test server or client side behavior.

The intent of the test is to verify the client gets an ABNORMAL_CLOSURE when sever initiated close while it was trying to send a `str`

**Comment by lenard-mosys** [comment]
> The intent of the test is to verify the client gets an ABNORMAL_CLOSURE when sever initiated close while it was trying to send a `str`

I think this is questionable. The server closed the connection cleanly, even if the client tried to send a message. There could be connection schemes where the client sends fire-and-forget status updates, and the server could close the connection whenever. The client should get the close code sent from the server, not abnormal closure in this case.

Having said that I struggle to find normative wording in the RFC that would describe the conforming behavior in this situation. The best I can find is (under "Normal Closure of Connections"):

https://datatracker.ietf.org/doc/html/rfc6455#section-7.3

> Servers MAY close the WebSocket connection whenever desired.

And for 1006:

>It is designated for use in applications expecting a status code to indicate that the connection was closed abnormally, e.g., without sending or receiving a Close control frame.

The "e.g." doesn't help here, suggesting that there are other situations where this is applicable.

Anyway, I think there are some problems with the autoclose mechanism of the client, and I can make a similar test fail to have the expected result, even when the client doesn't try to send a message that the server doesn't expect.

**Comment by bdraco** [comment]
Your changes seems reasonable to me. We likely need to adjust the test.  I agree the RFC is not crystal clear.

**Comment by lenard-mosys** [comment]
> Your changes seems reasonable to me. We likely need to adjust the test. I agree the RFC is not crystal clear.

In any case, there is probably a missing test that checks for the server side behavior. This particular test can be made to pass with `ws = web.WebSocketResponse(timeout=0.)`, which restores (roughly) the server-side behavior the client can test against, but I don't know how palatable that is.

**Comment by bdraco** [comment]
> > Your changes seems reasonable to me. We likely need to adjust the test. I agree the RFC is not crystal clear.
> 
> In any case, there is probably a missing test that checks for the server side behavior. This particular test can be made to pass with `ws = web.WebSocketResponse(timeout=0.)`, which restores (roughly) the server-side behavior the client can test against, but I don't know how palatable that is.

I think thats fine, but please add a comment to explain the 0 timeout.

**Comment by bdraco** [comment]
@lenard-mosys It looks like a conflict has appeared

File: CHANGES/9506.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Fixed ``WebSocketResponse.close()`` to discard non-close messages within its timeout window after sending close.
+-- by :user:`lenard-mosys`
```
Comment (bdraco) [inline CHANGES/9506.bugfix.rst]:
```suggestion
Fixed ``WebSocketResponse.close()`` to discard non-close messages within its timeout window after sending close -- by :user:`lenard-mosys`.
```

File: aiohttp/web_ws.py [hunk 2]
```diff
@@ -495,7 +495,11 @@ async def close(
 
         try:
             async with async_timeout.timeout(self._timeout):
-                msg = await reader.read()
+                while True:
+                    msg = await reader.read()
+                    if msg.type is WSMsgType.CLOSE:
```
Comment (bdraco) [inline aiohttp/web_ws.py:501]: We might get WSMsgType.ERROR as well

File: tests/test_web_websocket_functional.py [hunk 3]
```diff
@@ -1257,3 +1261,30 @@ async def handler(request: web.Request) -> web.WebSocketResponse:
...
+    close_code: Optional[WSCloseCode] = None
+
+    async def handler(request: web.Request) -> web.WebSocketResponse:
+        # Setting a short close timeout
+        ws = web.WebSocketResponse(timeout=0.1)
+        await ws.prepare(request)
+        await ws.close()
+
+        nonlocal close_code
+        if ws.close_code is not None:
```
Comment (bdraco) [inline tests/test_web_websocket_functional.py]: To fix the missing branch coverage, change this to an assert instead

File: aiohttp/web_ws.py (hunk 2 above)
Comment (lenard-mosys) [reply aiohttp/web_ws.py:501]: I reraise `msg.data` from here, so it's handled on the exception path.

Comment (bdraco) [reply aiohttp/web_ws.py:501]: I think that should be fine. Will need a test for this case as well

Comment (lenard-mosys) [reply aiohttp/web_ws.py:501]: Interestingly the `WSMsgType.ERROR` case seems to be already covered. The missing branch coverage is (probably) actually for the situation that this PR is meant to solve, which is when a regular message is received by the server within the close window. I added this test now.

**Comment by lenard-mosys** [comment]
> @lenard-mosys We are likely doing another release soon. Not sure how keen you are to get this in the next one, but just in case you missed [my comment above](https://github.com/aio-libs/aiohttp/pull/9508#discussion_r1824682943) about the missing branch coverage for https://github.com/aio-libs/aiohttp/pull/9508/files#diff-37008174f47abcf385c9bd15d713b586a921dc6aad7d75acd1510aebc75d2b4fR503

I'm not in a hurry with this. I added a test that hopefully covers the previously missing branch.

**Comment by bdraco** [comment]
Thanks. I'll take a look as soon as I can.

CI is a bit busy now as I'm cleaning up to prep another beta release

**Comment by bdraco** [comment]
Looks like that branch is still coming up on the coverage report

Screenshot 2024-11-05 at 11 13 16 AM

**Comment by bdraco** [comment]
https://app.codecov.io/gh/aio-libs/aiohttp/pull/9508/blob/aiohttp/web_ws.py

**Comment by lenard-mosys** [comment]
> Looks like that branch is still coming up on the coverage report

Weird. I don't understand how the next line is covered then. My added test should cover the hidden `else` case here. I guess I could just add an `else: pass` and see what happens. Anyway, I don't have more time today for this, I will follow up tomorrow.

**Comment by lenard-mosys** [comment]
I have some trouble covering the missing branch (that is producing a `WsMsgType.ERROR` message from `await reader.read()` within `WebSocketResponse.close()`). I tried sending an oversized message from the client side, with setting `WebSocketResponse(max_msg_size=8)` on the server size, but that produces an exception instead from `await reader.read()`.

It looks like that `WsMsgType.ERROR` is produced by exceptions, but those exceptions get reraised anyway from `await reader.read()`, therefore it's impossible to get this type of message at this level.

**Comment by bdraco** [comment]
It looks like the only place we could get one is from
```
    def _handle_ping_pong_exception(self, exc: BaseException) -> None:
        """Handle exceptions raised during ping/pong processing."""
        if self._closed:
            return
        self._set_closed()
        self._set_code_close_transport(WSCloseCode.ABNORMAL_CLOSURE)
        self._exception = exc
        if self._waiting and not self._closing and self._reader is not None:
            self._reader.feed_data(WSMessageError(data=exc, extra=None))
```

**Comment by bdraco** [comment]
For that to happen, close would have to be called right after `_handle_ping_pong_exception` happened

**Comment by bdraco** [comment]
So I agree its truly unreachable.

File: aiohttp/web_ws.py [hunk 4]
```diff
@@ -496,7 +496,13 @@ async def close(
...
+                        self._set_code_close_transport(msg.data)
+                        return True
+                    if msg.type is WSMsgType.ERROR:
+                        raise msg.data
```
Comment (bdraco) [inline aiohttp/web_ws.py]:
```suggestion
```

Sorry for sending you down a wild goose chase on this one.

**Review by bdraco — APPROVED** [review]

File: CHANGES/9506.bugfix.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+Fixed ``WebSocketResponse.close()`` to discard non-close messages within its timeout window after sending close -- by :user:`lenard-mosys`.
```
Comment (bdraco) [inline CHANGES/9506.bugfix.rst]:
```suggestion
Fixed :py:meth:`WebSocketResponse.close() <aiohttp.web.WebSocketResponse.close>` to discard non-close messages within its timeout window after sending close -- by :user:`lenard-mosys`.
```

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
3.10 https://github.com/aio-libs/aiohttp/pull/9687

**Comment by bdraco** [comment]
3.11 https://github.com/aio-libs/aiohttp/pull/9688

**Pull request merged**